### DISCUSSION TITLE
Firestore Translate Text Extension Integration

### DISCUSSION THREAD

**Discussion by IT-Support-AM** [flyerhq/flutter_chat_ui#457] [Ideas & Requests]
Hello,

I would like to ask if there is somebody who has figured out a way how to use the Firebase Extension "Translate Text in Firestore" (https://extensions.dev/extensions/firebase/firestore-translate-text) with the Chat UI Package. The use case is to translate the messages into the desired languages with the extension and then show the right message based on the location. 

Maybe someone already found a way to nicely integrate this extension. BTW, I'm also using the Flutter Chat Firebase Core package, and it would be good if the solution could integrate with that. 

I really like the project and want to say that I'm deeply grateful for such a nice, well done package.

Thanks in advance.

**Comment by demchenkoalex** [comment]
Hi!

I will be deprecating Firebase package, cause it is hard to maintain with all the requests, but if we are talking about chat UI specifically, I think it should work with this, looking at the doc:

```
This extension listens to your specified Cloud Firestore collection. If you add a string to a specified field in any document within that collection, this extension:

Translates the string into your specified target language(s); the source language of the string is automatically detected.
Adds the translation(s) of the string to a separate specified field in the same document.
```

basically as I understand when Firestore will translate text you can specify where it will write it, you write it somewhere, event comes back and you take the event and write that field to the text field of the text message. 

Trying to do v2 of the chat UI now, will try to include custom builders for all types of messages to handle cases if you will want to also switch between original/translated version.

**Reply by MonkeyDrone** [reply]
Deprecate firebase, does that mean v2 will no longer have firebase support?  I mean the package flutter firebase chat core.

**Reply by demchenkoalex** [reply]
there is no "support" of anything. The UI is backend agnostic, you just add an array of messages. It can come from anywhere, local storage, firebase, supabase, your backend, chatgpt, whatever. I created firebase chat core as an "example" implementation to show that yes, UI works with a backend. If you take a look in the code there, it just wraps a couple of calls to the Firebase through FlutterFire, super simple to do yourself. 

Also people say it is useless https://github.com/flyerhq/flutter_firebase_chat_core/issues/67#issuecomment-1581781701 and I tend to agree now

**Reply by IT-Support-AM** [reply]
Thanks for your reply!

Yeah i was just talking about the Flutter Chat UI but I also used the Firebase Core Package. I ended up saving the data from the translation package under the metadata field of a message document. This way I could access the properties within the message. Then I just used the TextMessageBuilder to create my own widget depending on which language is active for the user. If anyone is interested I paste the code below:

```dart
Widget messageBuilder(
    types.TextMessage message, {
    required int messageWidth,
    required bool showName,
  }) {
    final lc = Localizations.localeOf(context).languageCode;

    var translatedText = message.text;
    if (message.metadata != null) {
      switch (lc) {
        case 'de':
          translatedText = message.metadata!['de'] as String;
          break;
        case 'en':
          translatedText = message.metadata!['en'] as String;
          break;
        case 'pl':
          translatedText = message.metadata!['pl'] as String;
          break;
        case 'hr':
          translatedText = message.metadata!['hr'] as String;
          break;
      }
    }

    final newMessage = types.TextMessage(
      author: message.author,
      id: message.id,
      text: translatedText,
    );


   return TextMessage(
        emojiEnlargementBehavior: EmojiEnlargementBehavior.multi,
        hideBackgroundOnEmojiMessages: true,
        message: newMessage,
        showName: showName,
        usePreviewData: true,
   );
  }
```
